### PR TITLE
Upload shapefile on the frontend

### PR DESCRIPTION
**Pull request by amaury1093**
Small PR that:
- Adds an Upload Shapefile button
- Uploads the shapefile, converts it to geojson
- Show the geojson on the map, with the ability to edit it

Note:
- I tried a 100kb shapefile, the browser can handle it without problem.
- I tried with a 22mb shapefile, it's starting to get hard. On a 2016 macbook pro, there's a 10s lag. But eventually it showed the geojson, and it's editable.
- My guess is that the "editable" part is kind of hard on RAM. Maybe we could disable the editing if the uploaded shapefile is > xMB?

### CODE REVIEW

File: packages/app/components/EditMode/UploadButton/decorators/withUploadShapefile.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * Decorator to add a handler when we update the polygon we draw on the map.
+ */
+export default compose(
+  withEditMode,
+  withHandlers<WithEditModeProps, WithUploadShapefileProps>({
+    handleUploadShapefile: ({ updateGeojson }) => async ({
+      target: { files }
+    }: React.ChangeEvent<HTMLInputElement>) => {
+      try {
+        const uploaded = files[0];
```
Comment (quorth0n) [inline packages/app/components/EditMode/UploadButton/decorators/withUploadShapefile.ts]: Some validation should probably be performed here. A limit on the uploaded file size might be able to be configured on the web server. I added some client side validation to accept only .shp files, but there really should be some validation performed here as well to reject the request before we attempt to process it.

**Comment by quorth0n** [comment]
I noticed that after adding a territory with the upload shapefile button, adding another shapefile isn't possible. In other words it can only be done once per "edit view". Is this the intended behavior?

**Comment by quorth0n** [comment]
> My guess is that the "editable" part is kind of hard on RAM. Maybe we could disable the editing if the uploaded shapefile is > xMB?

Agree on this, maybe go with 10 mb for this limit?

**Comment by amaury1093** [comment]
So after some tests, even adding the Japan territory without editing it is kind of slow, but still doable (~5s on my machine).
- I decided to remove the edit on map feature completely for now. It's heavy on RAM, and I guess the mappers won't need to too soon, as they have already mapped on QGIS
- Yes, for now you can only upload 1 shapefile, uploading a new one replaces the previous one. I think mappers have 1 shapefile per PoliticalEntity? If needed, I can add support for uploading multiple shapefiles, shouldn't be too hard.

**Comment by amaury1093** [comment]
Re that last message, I manually deployed here: https://chronoscio-frontend-iivhzlcbpc.now.sh

File: packages/app/components/EditTerritory/UploadButton/decorators/withUploadShapefile.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * Decorator to add a handler when we update the polygon we draw on the map.
+ */
+export default compose(
+  subscribe('withEditTerritoryStore'),
+  withHandlers<WithEditTerritoryStoreProps, {}>({
+    handleUploadShapefile: ({ addShapefile }) => async ({
+      target: { files }
+    }: React.ChangeEvent<HTMLInputElement>) => {
+      try {
+        const uploaded = files[0];
```
Comment (quorth0n) [inline packages/app/components/EditTerritory/UploadButton/decorators/withUploadShapefile.ts:45]: Any logic to check mime type or file extension?

**Review by quorth0n — APPROVED** [review]
Otherwise LGTM

File: packages/app/components/EditTerritory/UploadButton/decorators/withUploadShapefile.ts (hunk 2 above)
Comment (amaury1093) [reply packages/app/components/EditTerritory/UploadButton/decorators/withUploadShapefile.ts:45]: Unfortunately the mime type is empty when I upload the shapefiles given by the mappers.

**Pull request merged**